pkg/storage/storage.go

```diff
@@ -80,69 +97,157 @@ func (p *storage) GetContainerMetrics(pods ...apitypes.NamespacedName) ([]api.Ti
 
-		contMetrics := make([]metrics.ContainerMetrics, len(metricPoint.Containers))
-		var earliestTS *time.Time
-		for i, contPoint := range metricPoint.Containers {
-			contMetrics[i] = metrics.ContainerMetrics{
+		var (
+			contMetrics = make([]metrics.ContainerMetrics, len(contPoints))
+			earliestTS  time.Time
+			window      time.Duration
+			contIdx     int
+		)
+		for contName, contPoint := range contPoints {
+			prevContPoint, present := p.prevPods[pod][contName]
+			if !present {
+				continue
+			}
+
+			cpuUsage := cpuUsageOverTime(contPoint.MetricsPoint, prevContPoint.MetricsPoint)
+			contMetrics[contIdx] = metrics.ContainerMetrics{
 				Name: contPoint.Name,
 				Usage: corev1.ResourceList{
-					corev1.ResourceName(corev1.ResourceCPU):    contPoint.CpuUsage,
+					corev1.ResourceName(corev1.ResourceCPU):    cpuUsage,
 					corev1.ResourceName(corev1.ResourceMemory): contPoint.MemoryUsage,
 				},
 			}
-			if earliestTS == nil || earliestTS.After(contPoint.Timestamp) {
-				ts := contPoint.Timestamp // copy to avoid loop iteration variable issues
-				earliestTS = &ts
+			if earliestTS.IsZero() || earliestTS.After(contPoint.Timestamp) {
+				earliestTS = contPoint.Timestamp
+				window = contPoint.Timestamp.Sub(prevContPoint.Timestamp)
 			}
+			contIdx++
 		}
-		if earliestTS == nil {
-			// we had no containers
-			earliestTS = &time.Time{}
-		}
-		timestamps[i] = api.TimeInfo{
-			Timestamp: *earliestTS,
-			Window:    kubernetesCadvisorWindow,
+		resMetrics[podIdx] = contMetrics
+		timestamps[podIdx] = api.TimeInfo{
+			Timestamp: earliestTS,
+			Window:    window,
 		}
-		resMetrics[i] = contMetrics
 	}
 	return timestamps, resMetrics, nil
 }
 
 func (p *storage) Store(batch *MetricsBatch) {
-	newNodes := make(map[string]NodeMetricsPoint, len(batch.Nodes))
-	var nodeCount, containerCount int
-	for _, nodePoint := range batch.Nodes {
+	p.storeNodeMetrics(batch.Nodes)
+	p.storePodMetrics(batch.Pods)
+}
+
+func (p *storage) storeNodeMetrics(nodes []NodeMetricsPoint) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	newNodes := make(map[string]NodeMetricsPoint, len(nodes))
+	prevNodes := make(map[string]NodeMetricsPoint, len(nodes))
+	var nodeCount int
+	for _, nodePoint := range nodes {
 		if _, exists := newNodes[nodePoint.Name]; exists {
 			klog.Errorf("duplicate node %s received", nodePoint.Name)
 			continue
 		}
-		nodeCount += 1
 		newNodes[nodePoint.Name] = nodePoint
+		nodeCount++
+
+		prevNodePoint, present := p.prevNodes[nodePoint.Name]
+		if present {
+			prevNodes[nodePoint.Name] = prevNodePoint
+		}
+
+		storedNodePoint, present := p.nodes[nodePoint.Name]
+		if !present {
+			continue
+		}
+
+		// If the new point is newer than the one stored for the container, move
+		// it to the list of the previous points.
+		// This check also prevents from updating the store if the same metric
+		// point was scraped twice.
+		if nodePoint.Timestamp.After(storedNodePoint.Timestamp) {
+			prevNodes[nodePoint.Name] = storedNodePoint
+		}
```

Comment:
Tried to rewrite it to be more readable. This way we have one block that sets value in prevNodes. Let's not use continue  to express `if else`. `Else` in Go is not used frequently as it promotes early exit in case of error, but this is not the case as both code branches are important.
```suggestion
		// If the new point is newer than the one stored for the container, move
		// stored point to the list of previous points. If not, copy the point previously stored point if available.
		// This prevents from storing same metric point twice.
               storedNodePoint, found := p.nodes[nodePoint.Name]
		if found && nodePoint.Timestamp.After(storedNodePoint.Timestamp) {
			prevNodes[nodePoint.Name] = storedNodePoint
		} else {
		  prevNodePoint, found := p.prevNodes[nodePoint.Name]
		  if found {
			prevNodes[nodePoint.Name] = prevNodePoint
		  }
		}
```

Reply:
sounds good, thank you for the suggestion! :slightly_smiling_face: